_posts/2021-04-14-glahn.md

```diff
@@ -0,0 +1,128 @@
+
+### Git-Remotes verwalten
+
+Git funktioniert deshalb so flexibel, weil jeder Klon eines Repositories wie ein eigenständiges Orginal funktioniert. Damit wir zusammenarbeiten können, müssen wir die Klone eines Git-Repositories noch verbinden. Das geschieht in der Regel über die Herkunft eines Klons, dem sog. `origin`. Diese Herkunft ist in unseren Git-Repositorien als `remote` abgelegt. Das können wir mit `git remote` und `git remote get-url origin` abfragen. 
+
+![git remote](/assets/images/post_20210413/git_remote.png)
+
+Ein Git-Repository kann aber auch mehr als ein Remote-Repository haben, damit wir Verbindungen zu mehreren Git-Servern halten können. Damit wir die Verbindungen auseinanderhalten können, sind die Verbindungen benannt. Diese Namen können wir uns ausdenken. Weil wir unser Repository umziehen wollen, ist ein guter Name `neworigin`.
+
+Wir erstellen eine neue Verbindung zu einem Git-Server (in unserem Fall der ZHAW GitHub Instanz) mit dem folgenden Kommando:
+
+```zsh
+git remote add neworigin https://github.zhaw.ch/dxiai/moodle-autopilot.git
+```
+
+Das spannende an diesem Befehl ist, dass die das entfernte Repository noch nicht existieren muss. Wichtig ist nur, dass ich ein neues Repository in dem angegebenen Bereich (`dxiai`) anlegen darf. 
+
+Weil es so schön ist, lege ich gleich noch ein Remote auf meinen *privaten* Git-Server an: 
+
+```zsh
+git remote add homeoffice https://git.0xff.li/phish108/moodle-autopilot.git
+```
+
+Das Ergebnis können wir wieder mit `git remote` abfragen. 
+
+![git multiple remotes](/assets/images/post_20210413/git_multiremote.png)
+
+Mein Repository hat jetzt 3 Remotes. 
+
+Bevor ich weitermache, sichere ich mich noch ab, dass ich auch alle Änderungen vom originalen Server geholt habe. Das erreiche ich mit den folgenden Kommandos:
+
+```zsh
+git fetch origin
+git pull origin master
+```
+
+### Ein Git-Repository auf einem Git-Server bereitstellen
+
+Mit diesen beiden neuen Remotes sind wir schon fast fertig mit dem Umzug. Jetzt muss nur noch das lokale Repository auf den entfernten Git-Server geschoben werden. Ein einfaches `git push` reicht hier nicht. Wir müssen zusätzlich noch die Option `--all` anfügen, damit Git das gesamte Repository mit allen Code-Zweigen auf die Server schiebt und ich keine Daten verliere. Ausserdem muss ich das Ziel der Aktion angeben, damit Git weiss, wohin die Daten sollen.
+
+![git push --all homeoffice](/assets/images/post_20210413/git_push_newremote.png)
+
+Diesen Schritt wiederhole ich für alle neuen Remotes. 
+
+### GitHub ist nicht (nur) Git!
+
+Beim Umziehen von einem Projekt auf GitHub oder auf einen GitHub-Enterprise-Server merken wir, dass GitHub nicht das gleiche wie Git ist, denn wir können nicht einfach mit `git push --all` ein Git-Repository zu GitHub schieben. Stattdessen erhalten wir einen Fehler.
```

Comment:
Bevor wir einen Fehler erhalten, werde ich direkt nach der Eingabe von git push --all nach Username for 'https://github.zhaw.ch' und nach 'Password for https://**bajk**@github.zhaw.ch' gefragt. 

Die Abfrage nach Username und Passwort geschieht sicher nicht nur bei mir. Denke, das sollte noch in den Blog aufgenommen werden.